# Ejercicio 71

Convertir una lista de Python en un arreglo tipo NumPy.

## Solución:

In [1]:
import numpy as np

In [2]:
lista = [1, 7, 11, 2, 19]
lista

[1, 7, 11, 2, 19]

In [3]:
type(lista)

list

In [4]:
arreglo = np.array(lista)
arreglo

array([ 1,  7, 11,  2, 19])

In [5]:
type(arreglo)

numpy.ndarray

In [6]:
np.array(list(range(0, 7)))

array([0, 1, 2, 3, 4, 5, 6])

# Ejercicio 72

Crear una matriz a partir de un arreglo que valores enteros arbitrarios.

## Solución:

In [7]:
import numpy as np

In [8]:
arreglo = np.arange(10, 22)
arreglo

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21])

In [9]:
matriz = arreglo.reshape(3, 4)
matriz

array([[10, 11, 12, 13],
       [14, 15, 16, 17],
       [18, 19, 20, 21]])

# Ejercicio 73

Crear un arreglo de 5 elementos, todos iguales a cero, y luego cambiar el último elemento por 2.

## Solución:

In [10]:
import numpy as np

In [11]:
arreglo = np.zeros(5)
arreglo

array([0., 0., 0., 0., 0.])

In [12]:
arreglo[-1] = 2
arreglo

array([0., 0., 0., 0., 2.])

In [13]:
arreglo[4] = 2
arreglo

array([0., 0., 0., 0., 2.])

# Ejercicio 74

Crear un arreglo a partir de un rango.

## Solución:

In [14]:
import numpy as np

In [15]:
arreglo1 = np.arange(1, 11)
arreglo1

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [16]:
arreglo2 = np.array(list(range(1, 11)))

In [17]:
arreglo2

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

# Ejercicio 75

Invertir los elementos de un arreglo.

## Solución:

In [18]:
import numpy as np

In [19]:
arreglo = np.arange(1, 11)
arreglo

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [20]:
arreglo[::-1]

array([10,  9,  8,  7,  6,  5,  4,  3,  2,  1])

# Ejercicio 76

Convertir los elementos de un arreglo a real (punto flotante, o _float_).

## Solución:

In [21]:
import numpy as np

In [22]:
lista = [2, 3, 5, 7, 11]
lista

[2, 3, 5, 7, 11]

In [23]:
arreglo = np.asfarray(lista)
arreglo

array([ 2.,  3.,  5.,  7., 11.])

In [24]:
arreglo.dtype

dtype('float64')

# Ejercicio 77

Crear un arreglo con elementos iguales a 1, y luego agregar un marco (o _padding_) con valores iguales a 0.

Entrada:<br>
1 1 1<br>
1 1 1<br>
1 1 1<br>

Salida:<br>
0 0 0 0 0<br>
0 1 1 1 0<br>
0 1 1 1 0<br>
0 1 1 1 0<br>
0 0 0 0 0<br>

## Solución:

In [25]:
import numpy as np

In [26]:
unos = np.ones((3, 3), dtype=np.int)
unos

array([[1, 1, 1],
       [1, 1, 1],
       [1, 1, 1]])

In [27]:
np.pad(unos, pad_width=1, mode='constant', constant_values=0)

array([[0, 0, 0, 0, 0],
       [0, 1, 1, 1, 0],
       [0, 1, 1, 1, 0],
       [0, 1, 1, 1, 0],
       [0, 0, 0, 0, 0]])

# Ejercicio 78

Crear una matriz que siga el patrón de distribución de celdas (casillas) de un tablero de ajedrez.

## Solución:

In [28]:
import numpy as np

In [29]:
tablero = np.zeros((8, 8), dtype=int)
tablero

array([[0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0]])

In [30]:
tablero[::2, 1::2] = 1
tablero

array([[0, 1, 0, 1, 0, 1, 0, 1],
       [0, 0, 0, 0, 0, 0, 0, 0],
       [0, 1, 0, 1, 0, 1, 0, 1],
       [0, 0, 0, 0, 0, 0, 0, 0],
       [0, 1, 0, 1, 0, 1, 0, 1],
       [0, 0, 0, 0, 0, 0, 0, 0],
       [0, 1, 0, 1, 0, 1, 0, 1],
       [0, 0, 0, 0, 0, 0, 0, 0]])

In [31]:
tablero[1::2, ::2] = 1
tablero

array([[0, 1, 0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0, 1, 0],
       [0, 1, 0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0, 1, 0],
       [0, 1, 0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0, 1, 0],
       [0, 1, 0, 1, 0, 1, 0, 1],
       [1, 0, 1, 0, 1, 0, 1, 0]])

# Ejercicio 79

Crear una matriz a partir de una tupla de listas.

## Solución:

In [32]:
import numpy as np

In [33]:
tupla = ([1, 2, 3], [4, 5, 6], [7, 8, 9])
tupla

([1, 2, 3], [4, 5, 6], [7, 8, 9])

In [34]:
np.asarray(tupla)

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

# Ejercicio 80

Agregar elementos al final de un arreglo que ya contiene valores arbitrarios.

## Solución:

In [35]:
import numpy as np

In [36]:
arreglo = np.array([2, 3, 5, 7])
arreglo

array([2, 3, 5, 7])

In [37]:
arreglo = np.append(arreglo, [11, 13, 17])
arreglo

array([ 2,  3,  5,  7, 11, 13, 17])

# Ejercicio 81

Crear un arreglo vacío, y otro con datos.

## Solución:

In [38]:
import numpy as np

In [39]:
matriz = np.empty((4, 4))
matriz

array([[4.67296746e-307, 1.69121096e-306, 1.06811083e-306,
        8.34441742e-308],
       [1.78022342e-306, 6.23058028e-307, 9.79107872e-307,
        6.89807188e-307],
       [7.56594375e-307, 6.23060065e-307, 1.78021527e-306,
        8.34454050e-308],
       [1.11261027e-306, 1.15706896e-306, 1.33512173e-306,
        1.33504432e-306]])

In [40]:
arreglo = np.full((4, 4), 5)
arreglo

array([[5, 5, 5, 5],
       [5, 5, 5, 5],
       [5, 5, 5, 5],
       [5, 5, 5, 5]])

# Ejercicio 82

Convertir a grados Fahrenheit cada uno de los elementos de un arreglo que contiene medidas en grados centigrados.

# Solución:

In [41]:
import numpy as np

In [42]:
centigrados = np.array([0, 32.5, 42.8, 13.9])
centigrados

array([ 0. , 32.5, 42.8, 13.9])

C*9/5+32

In [43]:
centigrados*9/5 + 32

array([ 32.  ,  90.5 , 109.04,  57.02])

# Ejercicio 83

Obtener la parte real e imaginaria de un arreglo que contiene números complejos.

## Solución:

In [44]:
import numpy as np

In [45]:
arreglo = np.sqrt([3+8j])
arreglo

array([2.40249909+1.66493299j])

In [46]:
arreglo.real

array([2.40249909])

In [47]:
arreglo.imag

array([1.66493299])

# Ejercicio 84

Contar la cantidad de elementos de un arreglo, el espacio en bytes que ocupa cada uno de ellos, y el tamaño total en bytes que ocupa todo el arreglo.

## Solución:

In [48]:
import numpy as np

In [49]:
arreglo = np.array([2, 3, 5, 7], dtype=np.int64)
arreglo

array([2, 3, 5, 7], dtype=int64)

In [50]:
arreglo.size

4

In [51]:
arreglo.itemsize

8

In [52]:
arreglo.nbytes

32

# Ejercicio 85

Comprobar si cada elemento de un arreglo está contenido en un segundo arreglo.

## Solución:

In [53]:
import numpy as np

In [54]:
arreglo1 = np.array([2, 3, 5, 7, 11, 13])
arreglo1

array([ 2,  3,  5,  7, 11, 13])

In [55]:
arreglo2 = np.array([2, 11])

In [56]:
arreglo2

array([ 2, 11])

In [57]:
np.in1d(arreglo1, arreglo2)

array([ True, False, False, False,  True, False])

# Ejercicio 86

Encontrar los elementos comunes entre dos arreglos.

## Solución:

In [58]:
import numpy as np

In [59]:
arreglo1 = np.array([2, 3, 7, 19, 23])
arreglo1

array([ 2,  3,  7, 19, 23])

In [60]:
arreglo2 = np.array([7, 23, 2, 8])
arreglo2

array([ 7, 23,  2,  8])

In [61]:
np.intersect1d(arreglo1, arreglo2)

array([ 2,  7, 23])

# Ejercicio 87

Obtener los elementos únicos de un arreglo.

## Solución:

In [62]:
import numpy as np

In [63]:
arreglo = np.array([23, 11, 9, 8, 19, 8, 11, 23, 1, 14])
arreglo

array([23, 11,  9,  8, 19,  8, 11, 23,  1, 14])

In [64]:
np.unique(arreglo)

array([ 1,  8,  9, 11, 14, 19, 23])

In [65]:
matriz = np.array([[1, 2, 3], [3, 4, 1]])
matriz

array([[1, 2, 3],
       [3, 4, 1]])

In [66]:
np.unique(matriz)

array([1, 2, 3, 4])

# Ejercicio 88

Encontrar la diferencia entre dos arreglos.

## Solución:

In [67]:
 import numpy as np

In [68]:
arregloA = np.array([2, 3, 7, 0, -5, -7, 11, 16])
arregloA

array([ 2,  3,  7,  0, -5, -7, 11, 16])

In [69]:
arregloB = np.array([7, 9, 15, 18, 0, -1])
arregloB

array([ 7,  9, 15, 18,  0, -1])

In [70]:
np.setdiff1d(arregloA, arregloB)

array([-7, -5,  2,  3, 11, 16])

# Ejercicio 89

Obtener los elementos únicos de dos arreglos.

## Solución:

In [71]:
import numpy as np

In [72]:
arregloA = np.array([1, 2, 3, 4])
arregloA

array([1, 2, 3, 4])

In [73]:
arregloB = np.array([3, 4, 5, 6])
arregloB

array([3, 4, 5, 6])

In [74]:
np.setxor1d(arregloA, arregloB)

array([1, 2, 5, 6])

# Ejercicio 90

Encontrar la unión de dos arreglos.

## Solución:

In [75]:
import numpy as np

In [76]:
arregloA = np.array([1, 2, 3, 4])

In [77]:
arregloA

array([1, 2, 3, 4])

In [78]:
arregloB = np.array([4, 5, 6, 7])
arregloB

array([4, 5, 6, 7])

In [79]:
np.union1d(arregloA, arregloB)

array([1, 2, 3, 4, 5, 6, 7])

# Ejercicio 91

Comprobar si todos los elementos de un arreglo tiene valor de verdad igual a `True`.

## Solución:

In [80]:
import numpy as np

In [81]:
arreglo = np.array([True, True, False, True])
arreglo

array([ True,  True, False,  True])

In [82]:
np.all(arreglo)

False

In [83]:
arreglo = np.array([10, 30, 95, 0, 13])
arreglo

array([10, 30, 95,  0, 13])

In [84]:
np.all(arreglo)

False

In [85]:
np.all(np.array([True, True, True]))

True

# Ejercicio 92

Crear un nuevo arreglo a partir de la repitición de los elementos de uno existente.

## Solución:

In [86]:
import numpy as np

In [87]:
arreglo = np.array([1, 2, 3])
arreglo

array([1, 2, 3])

In [88]:
arreglo_nuevo = np.tile(arreglo, 3)
arreglo_nuevo

array([1, 2, 3, 1, 2, 3, 1, 2, 3])

In [89]:
np.tile(arreglo, 5)

array([1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3])

# Ejercicio 93

Crear un arreglo a partir de la repetición de un elmento.

## Solución:

In [90]:
import numpy as np

In [91]:
arreglo = np.repeat(5, 4)
arreglo

array([5, 5, 5, 5])

In [92]:
np.repeat(np.array([[1, 2], [5, 7]]), 3)

array([1, 1, 1, 2, 2, 2, 5, 5, 5, 7, 7, 7])

# Ejercicio 94

Encontrar los índices de los elementos menor y mayor en un arreglo con valores arbitrarios.

## Solución:

In [93]:
import numpy as np

In [94]:
arreglo = np.array([7, 13, 9, 0, 8, 19, 14, 21, 1, 29])
arreglo

array([ 7, 13,  9,  0,  8, 19, 14, 21,  1, 29])

In [95]:
np.argmin(arreglo)

3

In [96]:
np.argmax(arreglo)

9

# Ejercicio 95

Comparar a través de operadores relacionales los elementos de dos arreglos.

## Solución:

In [97]:
import numpy as np

`>` : np.greater(a, b)<br>
`>=` : np.greater_equal(a, b)<br>
`<` : np.less(a, b)<br>
`<=` : np.less_equal(a, b)

In [98]:
a = np.array([1, 3, 7, 9])
b = np.array([7, 14, -1, 8])

In [99]:
np.greater(a, b)

array([False, False,  True,  True])

In [100]:
np.greater_equal(a, b)

array([False, False,  True,  True])

In [101]:
np.less(a, b)

array([ True,  True, False, False])

In [102]:
np.less_equal(a, b)

array([ True,  True, False, False])

# Ejercicio 96

Ordenar una matriz por medio del eje de las filas y también por el eje de las columnas.

## Solución:

In [103]:
import numpy as np

In [104]:
matriz = np.array([[4, 3], [2, 1]])
matriz

array([[4, 3],
       [2, 1]])

In [105]:
np.sort(matriz, axis=0)

array([[2, 1],
       [4, 3]])

In [106]:
np.sort(matriz, axis=1)

array([[3, 4],
       [1, 2]])

# Ejercicio 97

Ordenar un grupo de apellidos y nombres y obtener los índices del ordenamiento resultante.

## Solución:

In [107]:
import numpy as np

In [108]:
nombres = ['Edward', 'Angela', 'Daniela', 'German']
apellidos = ['Ortiz', 'Burgos', 'Meneses', 'Urbano']

np.lexsort((nombres, apellidos))

array([1, 2, 0, 3], dtype=int64)

# Ejercicio 98

Crear un filtro sobre los elementos de un arreglo.

## Solución:

In [109]:
import numpy as np

In [110]:
arreglo = np.array([13, 18, 9, 3, 7, 15, 23])
arreglo

array([13, 18,  9,  3,  7, 15, 23])

In [111]:
arreglo[arreglo > 9]

array([13, 18, 15, 23])

# Ejercicio 99

Cambiar la dimensión de un arreglo.

## Solución:

In [112]:
import numpy as np

In [113]:
arreglo = np.arange(1, 13)
arreglo

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

In [114]:
arreglo.shape

(12,)

In [115]:
arreglo.shape = (3, 4)

In [116]:
arreglo

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

# Ejercicio 100

Convertir una matriz en un arreglo.

## Solución:

In [117]:
import numpy as np

In [118]:
matriz = np.arange(1, 13).reshape(3, 4)

matriz

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

In [119]:
matriz.shape

(3, 4)

In [120]:
np.ravel(matriz)

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

# Ejercicio 101

Cambiar el tipo de dato de un arreglo.

## Solución:

In [121]:
import numpy as np

In [122]:
x = np.array([1, 2, 3, 4, 5], np.int32)
x

array([1, 2, 3, 4, 5])

In [123]:
x.dtype

dtype('int32')

In [124]:
y = x.astype(float)
y

array([1., 2., 3., 4., 5.])

In [125]:
y.dtype

dtype('float64')

# Ejercicio 102

Crear una matriz de 5 filas y 4 columnas con elementos iguales a 3.

## Solución:

In [126]:
import numpy as np

In [127]:
matriz = np.ones([5, 4], dtype=np.int)

matriz

array([[1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1]])

In [128]:
matriz * 3

array([[3, 3, 3, 3],
       [3, 3, 3, 3],
       [3, 3, 3, 3],
       [3, 3, 3, 3],
       [3, 3, 3, 3]])

# Ejercicio 103

Crear un arreglo con dimensión y tipo de dato a otro arreglo. Se debe cambiar el contenido por otro valor.

## Solución:

In [129]:
import numpy as np

x = np.arange(5, dtype=np.int64)
x

array([0, 1, 2, 3, 4], dtype=int64)

In [130]:
y = np.full_like(x, 7)

y

array([7, 7, 7, 7, 7], dtype=int64)

# Ejercicio 104

Crear una matriz con valores arbitrarios en la diagonal principal; el resto de elementos deben ser igual a 0.

## Solución:

In [131]:
import numpy as np

In [132]:
matriz = np.diagflat([2, 4, 6, 8, 9])

matriz

array([[2, 0, 0, 0, 0],
       [0, 4, 0, 0, 0],
       [0, 0, 6, 0, 0],
       [0, 0, 0, 8, 0],
       [0, 0, 0, 0, 9]])

# Ejercicio 105

Crear un rango de valores enteros como un arreglo NumPy.

## Solución:

In [133]:
import numpy as np

In [134]:
arreglo = np.arange(20, 41)

arreglo

array([20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36,
       37, 38, 39, 40])

# Ejercicio 106

Crear un arreglo con 20 elementos entre 0.0 y 1. Deben estar separados o espaciados uniformemente.

## Solución:

In [135]:
import numpy as np

In [136]:
arreglo = np.linspace(0.0, 1.0, 20)

arreglo

array([0.        , 0.05263158, 0.10526316, 0.15789474, 0.21052632,
       0.26315789, 0.31578947, 0.36842105, 0.42105263, 0.47368421,
       0.52631579, 0.57894737, 0.63157895, 0.68421053, 0.73684211,
       0.78947368, 0.84210526, 0.89473684, 0.94736842, 1.        ])

# Ejercicio 107

Crear un rango de 10 elementos en una escala logarítmica. El punto inicial debe ser 2 y el final 5.

## Solución:

In [137]:
import numpy as np

In [138]:
arreglo = np.logspace(2, 5, 10, endpoint=False)

arreglo

array([  100.        ,   199.5262315 ,   398.10717055,   794.32823472,
        1584.89319246,  3162.27766017,  6309.5734448 , 12589.25411794,
       25118.8643151 , 50118.72336273])

# Ejercicio 108

Crear una matriz con valores iguales a uno a partir de una diagonal arbitraria.

## Solución:

In [139]:
import numpy as np

In [140]:
matriz = np.tri(5, 4, -1)

matriz

array([[0., 0., 0., 0.],
       [1., 0., 0., 0.],
       [1., 1., 0., 0.],
       [1., 1., 1., 0.],
       [1., 1., 1., 1.]])

# Ejercicio 109

Crear una matriz con valores arbitrarios. Luego, la matriz debe tener elementos iguales a cero por debajo de una diagonal.

## Solución:

In [141]:
import numpy as np

In [142]:
matriz = np.triu(np.arange(1, 13).reshape(4, 3), 0)

matriz

array([[1, 2, 3],
       [0, 5, 6],
       [0, 0, 9],
       [0, 0, 0]])

# Ejercicio 110

Encontrar el n-ésimo elemento en una matriz.

## Solución:

In [143]:
import numpy as np

In [144]:
matriz = np.array([[7, 13, 7, 5], [3, 8, 4, 2]])

matriz

array([[ 7, 13,  7,  5],
       [ 3,  8,  4,  2]])

In [145]:
matriz.flat[4]

3

# Ejercicio 111

Concatenar dos matrices.

## Solución:

In [146]:
import numpy as np

In [147]:
matriz_a = np.arange(1, 5).reshape(2, 2)

matriz_a

array([[1, 2],
       [3, 4]])

In [148]:
matriz_b = np.arange(5, 9).reshape(2, 2)

matriz_b

array([[5, 6],
       [7, 8]])

In [149]:
matriz_c = np.concatenate((matriz_a, matriz_b), 0)

matriz_c

array([[1, 2],
       [3, 4],
       [5, 6],
       [7, 8]])

# Ejercicio 112

Apilar dos arreglos como columnas.

## Solución:

In [150]:
import numpy as np

In [151]:
arreglo_a = np.array([1, 2, 3])
arreglo_b = np.array([4, 5, 6])

matriz = np.column_stack((arreglo_a, arreglo_b))

In [152]:
matriz

array([[1, 4],
       [2, 5],
       [3, 6]])

# Ejercicio 113

Particionar un arreglo en cuatro (4) subarreglos.

## Solución:

In [153]:
import numpy as np

In [154]:
arreglo = np.arange(1, 20)

arreglo

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19])

In [155]:
np.split(arreglo, [4, 10, 15])

[array([1, 2, 3, 4]),
 array([ 5,  6,  7,  8,  9, 10]),
 array([11, 12, 13, 14, 15]),
 array([16, 17, 18, 19])]

# Ejercicio 114

Particionar una matriz a través del eje de las columnas.

## Solución:

In [156]:
import numpy as np

In [157]:
matriz = np.arange(1, 26).reshape((5, 5))
matriz

array([[ 1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10],
       [11, 12, 13, 14, 15],
       [16, 17, 18, 19, 20],
       [21, 22, 23, 24, 25]])

In [158]:
np.hsplit(matriz, [2, 4])

[array([[ 1,  2],
        [ 6,  7],
        [11, 12],
        [16, 17],
        [21, 22]]), array([[ 3,  4],
        [ 8,  9],
        [13, 14],
        [18, 19],
        [23, 24]]), array([[ 5],
        [10],
        [15],
        [20],
        [25]])]

# Ejercicio 115

Encontrar en un arreglo la cantidad de elementos distintos que son distintos de cero (0).

## Solución:

In [159]:
import numpy as np

In [160]:
arreglo = np.array([[5, 7, 0, 3], [9, 13, -8, 14], [0, 0, 15, 19]])

arreglo

array([[ 5,  7,  0,  3],
       [ 9, 13, -8, 14],
       [ 0,  0, 15, 19]])

In [161]:
np.count_nonzero(arreglo)

9

# Ejercicio 116

Crear una matriz de 4x4 y por cada fila tener valores entre 3 y 6.

## Solución:

In [162]:
import numpy as np

In [163]:
matriz = np.zeros((4, 4))

matriz

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [164]:
matriz += np.arange(3, 7)

In [165]:
matriz

array([[3., 4., 5., 6.],
       [3., 4., 5., 6.],
       [3., 4., 5., 6.],
       [3., 4., 5., 6.]])

# Ejercicio 117

Comprobar si un valor arbitrario se encuentra de los valores almacenados en un arreglo.

## Solución:

In [166]:
import numpy as np

In [167]:
matriz = np.array([[1, 14, 17.5, 9], [9, 7, 4, 13], [1, 2, 0.5, 3]])

matriz

array([[ 1. , 14. , 17.5,  9. ],
       [ 9. ,  7. ,  4. , 13. ],
       [ 1. ,  2. ,  0.5,  3. ]])

In [168]:
# in

7 in matriz

True

In [169]:
7.9 in matriz

False

In [170]:
0 in matriz

False

# Ejercicio 118

Crear un arreglo con 20 valores uniformemente espaciados 0 y 1 (excluídos los extremos).

## Solución:

In [171]:
import numpy as np

In [172]:
arreglo = np.linspace(0, 1, 22, endpoint=True)

arreglo

array([0.        , 0.04761905, 0.0952381 , 0.14285714, 0.19047619,
       0.23809524, 0.28571429, 0.33333333, 0.38095238, 0.42857143,
       0.47619048, 0.52380952, 0.57142857, 0.61904762, 0.66666667,
       0.71428571, 0.76190476, 0.80952381, 0.85714286, 0.9047619 ,
       0.95238095, 1.        ])

In [173]:
arreglo = arreglo[1:-1]

arreglo

array([0.04761905, 0.0952381 , 0.14285714, 0.19047619, 0.23809524,
       0.28571429, 0.33333333, 0.38095238, 0.42857143, 0.47619048,
       0.52380952, 0.57142857, 0.61904762, 0.66666667, 0.71428571,
       0.76190476, 0.80952381, 0.85714286, 0.9047619 , 0.95238095])

# Ejercicio 119

Especificar que un arreglo sea de sólo lectura (no modificable).

## Solución:

In [174]:
import numpy as np

In [175]:
arreglo = np.array([1, 2, 3, 4, 5])

arreglo

array([1, 2, 3, 4, 5])

In [176]:
arreglo[-1] = 6

arreglo

array([1, 2, 3, 4, 6])

In [177]:
arreglo.flags.writeable = False

In [178]:
arreglo[0] = 8

ValueError: assignment destination is read-only

# Ejercicio 120

Encontrar los elementos que son múltiplo de 2 y 3 en un arreglo; luego calcular su suma.

## Solución:

In [179]:
import numpy as np

In [180]:
arreglo = np.arange(1, 101)

arreglo

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99, 100])

In [181]:
arreglo = arreglo[(arreglo % 2 == 0) | (arreglo % 3 == 0)]

arreglo

array([  2,   3,   4,   6,   8,   9,  10,  12,  14,  15,  16,  18,  20,
        21,  22,  24,  26,  27,  28,  30,  32,  33,  34,  36,  38,  39,
        40,  42,  44,  45,  46,  48,  50,  51,  52,  54,  56,  57,  58,
        60,  62,  63,  64,  66,  68,  69,  70,  72,  74,  75,  76,  78,
        80,  81,  82,  84,  86,  87,  88,  90,  92,  93,  94,  96,  98,
        99, 100])

In [182]:
arreglo.sum()

3417

# Ejercicio 121

Crear un arreglo con 10000 elementos.

## Solución:

In [183]:
import numpy as np

In [184]:
arreglo = np.arange(1e4)

arreglo

array([0.000e+00, 1.000e+00, 2.000e+00, ..., 9.997e+03, 9.998e+03,
       9.999e+03])

In [185]:
len(arreglo)

10000

# Ejercicio 122

Iterar a través de cada elemento de un arreglo o matriz.

## Solución:

In [186]:
import numpy as np

In [187]:
matriz = np.arange(1, 16).reshape(3, 5)

matriz

array([[ 1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10],
       [11, 12, 13, 14, 15]])

In [188]:
for elemento in np.nditer(matriz):
    print(elemento, end=' ')

1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 

# Ejercicio 123

Iterar una matriz siguiendo el mecanismo de recorrido basado en Fortran.

## Solución:

In [189]:
import numpy as np

In [190]:
matriz = np.arange(1, 13).reshape(4, 3)

matriz

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [191]:
for elemento in np.nditer(matriz, order='F'):
    print(elemento, end=' ')

1 4 7 10 2 5 8 11 3 6 9 12 

# Ejercicio 124

Iterar una matriz y mutiplicar cada elemento por un valor arbitrario.

## Solución:

In [192]:
import numpy as np

In [193]:
matriz = np.arange(16).reshape(4, 4)

matriz

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15]])

In [194]:
for elemento in np.nditer(matriz, op_flags=['readwrite']):
    elemento[...] = 7 * elemento
    
matriz

array([[  0,   7,  14,  21],
       [ 28,  35,  42,  49],
       [ 56,  63,  70,  77],
       [ 84,  91,  98, 105]])

# Ejercicio 125

Combinar un arreglo con una matriz.

## Solución:

In [195]:
import numpy as np

In [196]:
arreglo = np.arange(1, 6)
arreglo

array([1, 2, 3, 4, 5])

In [197]:
matriz = np.arange(5, 20).reshape(3, 5)

matriz

array([[ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19]])

In [198]:
for i, j in np.nditer([arreglo, matriz]):
    print('{}:{}'.format(i, j))

1:5
2:6
3:7
4:8
5:9
1:10
2:11
3:12
4:13
5:14
1:15
2:16
3:17
4:18
5:19


# Ejercicio 126

Crear una matriz con columnas de diferente tipo de dato.

## Solución:

In [199]:
import numpy as np

In [200]:
matriz = np.zeros((3, ), dtype=('i4,f4,a64'))

In [201]:
datos = [(1, 4.1, 'Cronica Roja'), (2, 4.3, 'La virgen de los sicarios'), (3, 4.2, 'Satanas')]

datos

[(1, 4.1, 'Cronica Roja'),
 (2, 4.3, 'La virgen de los sicarios'),
 (3, 4.2, 'Satanas')]

In [202]:
matriz[:] = datos

matriz

array([(1, 4.1, b'Cronica Roja'), (2, 4.3, b'La virgen de los sicarios'),
       (3, 4.2, b'Satanas')],
      dtype=[('f0', '<i4'), ('f1', '<f4'), ('f2', 'S64')])

# Ejercicio 127

A través de una función elevar al cubo cada elemento de un arreglo.

## Solución:

In [203]:
import numpy as np

In [204]:
def elevar_cubo(arreglo):
    iterador = np.nditer([arreglo, None])
    
    for a, b in iterador:
        b[...] = a * a * a
        
    return iterador.operands[1]

In [205]:
elevar_cubo([3, 5, 7])

array([ 27, 125, 343])

# Ejercicio 128

Convertir una matriz en una lista Python.

## Solución:

In [206]:
import numpy as np

In [207]:
matriz = np.arange(1, 13).reshape(3, 4)
matriz

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

In [208]:
matriz.tolist()

[[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]

# Ejercicio 129

Obtener cada fila de una matriz.

## Solución:

In [209]:
import numpy as np

In [210]:
matriz = np.arange(0, 12).reshape(3, 4)

matriz

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [211]:
for i in range(len(matriz)):
    print(matriz[i,:])

[0 1 2 3]
[4 5 6 7]
[ 8  9 10 11]


# Ejercicio 130

Obtener cada columna de una matriz.

## Solución:

In [212]:
import numpy as np

In [213]:
matriz = np.arange(1, 13).reshape(4, 3)

matriz

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [214]:
for c in range(len(matriz[1,:])):
    print(matriz[:, c])

[ 1  4  7 10]
[ 2  5  8 11]
[ 3  6  9 12]


# Ejercicio 131

Convertir un arreglo con valores reales en uno con valores enteros.

## Solución:

In [215]:
import numpy as np

In [216]:
arreglo = np.arange(1, 9, dtype=np.float)

arreglo

array([1., 2., 3., 4., 5., 6., 7., 8.])

In [217]:
arreglo = arreglo.astype(np.int)

arreglo

array([1, 2, 3, 4, 5, 6, 7, 8])

# Ejercicio 132

Imprimir el contenido de un arreglo con valores reales con precisión arbitraria.

## Solución:

In [218]:
import numpy as np

In [219]:
arreglo = np.random.rand(10)

arreglo

array([0.72684695, 0.03508413, 0.1592384 , 0.00170631, 0.10910179,
       0.04879539, 0.22848149, 0.47913011, 0.50359785, 0.17056554])

In [220]:
np.set_printoptions(precision=5)

In [221]:
arreglo

array([0.72685, 0.03508, 0.15924, 0.00171, 0.1091 , 0.0488 , 0.22848,
       0.47913, 0.5036 , 0.17057])

# Ejercicio 133

Deshabilitar el uso de notación científica para valores reales pequeños.

## Solución:

In [222]:
import numpy as np

In [223]:
arreglo = np.array([2.6e-10, 1.6, 1500, .235])

arreglo

array([2.60e-10, 1.60e+00, 1.50e+03, 2.35e-01])

In [224]:
np.set_printoptions(suppress=True)

In [225]:
arreglo

array([   0.   ,    1.6  , 1500.   ,    0.235])

# Ejercicio 134

Crear un arreglo a partir de un objeto generador.

## Solución:

In [226]:
import numpy as np

In [227]:
generador = (elemento for elemento in range(10))

In [228]:
type(generador)

generator

In [229]:
arreglo = np.fromiter(generador, np.int)

In [230]:
arreglo

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

# Ejercicio 135

Agregar una columna adicional a una matriz.

## Solución:

In [231]:
import numpy as np

In [232]:
matriz = np.array([[1, 2, 3, 4], [5, 6, 7, 8]])

matriz

array([[1, 2, 3, 4],
       [5, 6, 7, 8]])

In [233]:
arreglo = np.array([[9], [10]])

arreglo

array([[ 9],
       [10]])

In [234]:
matriz = np.append(matriz, arreglo, axis=1)

matriz

array([[ 1,  2,  3,  4,  9],
       [ 5,  6,  7,  8, 10]])

# Ejercicio 136

Reemplazar los valores que cumplan una condición particular por un valor arbitrario.

## Solución:

In [235]:
import numpy as np

In [236]:
matriz = np.array([[8, 9, 13, 14], [13, -5, 2, 0], [8, 14, 1, 3]])

matriz

array([[ 8,  9, 13, 14],
       [13, -5,  2,  0],
       [ 8, 14,  1,  3]])

In [237]:
matriz[matriz > 10] = 10

matriz

array([[ 8,  9, 10, 10],
       [10, -5,  2,  0],
       [ 8, 10,  1,  3]])

# Ejercicio 137

Remover elementos de un arreglo a partir de sus índices.

## Solución:

In [238]:
import numpy as np

In [239]:
arreglo = np.array([7, 4, 5, 9, 13, 18, 5, 14])

arreglo

array([ 7,  4,  5,  9, 13, 18,  5, 14])

In [240]:
indices = [4, 0, 1]

In [241]:
arreglo_nuevo = np.delete(arreglo, indices)

arreglo_nuevo

array([ 5,  9, 18,  5, 14])

# Ejercicio 138

Reemplazar los valores negativos o iguales a cero por el valor cero (0).

## Solución:

In [242]:
import numpy as np

In [243]:
arreglo = np.array([-5, 8, 14, 13, 0, 8, -1, -2])

arreglo

array([-5,  8, 14, 13,  0,  8, -1, -2])

In [244]:
arreglo[arreglo <= 0] = 0

arreglo

array([ 0,  8, 14, 13,  0,  8,  0,  0])

# Ejercicio 139

Remover las filas que contengan valores no numéricos.

## Solución:

In [245]:
import numpy as np

In [246]:
m = np.array([[1, 2, 3, 4], [8, 9, 7, 3], [np.nan, np.nan, 8, 7]])

m

array([[ 1.,  2.,  3.,  4.],
       [ 8.,  9.,  7.,  3.],
       [nan, nan,  8.,  7.]])

In [247]:
m = m[~np.isnan(m).any(axis=1)]

m

array([[1., 2., 3., 4.],
       [8., 9., 7., 3.]])

# Ejercicio 140

Imprimir el contenido de un arreglo.

## Solución:

In [248]:
import numpy as np

In [249]:
arreglo = np.array([1, 2, 3, 4, 5, 6])

arreglo

array([1, 2, 3, 4, 5, 6])

In [250]:
print(arreglo)

[1 2 3 4 5 6]


# Ejercicio 141

Convertir un arreglo en una cadena binaria, y está última reestablecerla en el arreglo original.

## Solución:

In [251]:
import numpy as np

In [252]:
arreglo = np.array([2, 3, 5, 7, 11], float)

arreglo

array([ 2.,  3.,  5.,  7., 11.])

In [253]:
binaria = arreglo.tostring()

binaria

b'\x00\x00\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00\x00\x08@\x00\x00\x00\x00\x00\x00\x14@\x00\x00\x00\x00\x00\x00\x1c@\x00\x00\x00\x00\x00\x00&@'

In [254]:
arreglo2 = np.fromstring(binaria)

arreglo2

C:\Users\johno\Anaconda3-2018.12\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: The binary mode of fromstring is deprecated, as it behaves surprisingly on unicode inputs. Use frombuffer instead
  """Entry point for launching an IPython kernel.


array([ 2.,  3.,  5.,  7., 11.])

# Ejercicio 142

Calcular la sumatoria y la productoria del contenido de un arreglo.

## Solución:

In [255]:
import numpy as np

In [256]:
a = np.array([2, 3, 5, 7, 11])

a

array([ 2,  3,  5,  7, 11])

In [257]:
a.sum()

28

In [258]:
a.prod()

2310

# Ejercicio 143

Calcular la magnitud de un arreglo.

## Solución:

In [259]:
import numpy as np

In [260]:
arreglo = np.array([2, 7, 13, 5, 3])

arreglo

array([ 2,  7, 13,  5,  3])

In [261]:
np.linalg.norm(arreglo)

16.0

# Ejercicio 144

Obtener los elementos únicos de un arreglo, y además sus frecuencias.

## Solución:

In [262]:
import numpy as np

In [263]:
a = np.array([2, 3, 5, 3, 3, 5, 2, 7, 9, 5, 9])

a

array([2, 3, 5, 3, 3, 5, 2, 7, 9, 5, 9])

In [264]:
unicos, conteos = np.unique(a, return_counts=True)

In [265]:
unicos, conteos

(array([2, 3, 5, 7, 9]), array([2, 3, 3, 1, 2], dtype=int64))

# Ejercicio 145

Comprobar si un arreglo está vacío.

## Solución:

In [266]:
import numpy as np

In [267]:
a = np.array([2, 3])
b = np.array([])

a

array([2, 3])

In [268]:
b

array([], dtype=float64)

In [269]:
a.size

2

In [270]:
b.size

0

# Ejercicio 146

Dividir cada fila de una matriz por un arreglo.

## Solución:

In [271]:
import numpy as np

In [272]:
m = np.array([[2, 3, 4], [4, 6, 8], [8, 9, 16]])

m

array([[ 2,  3,  4],
       [ 4,  6,  8],
       [ 8,  9, 16]])

In [273]:
a = np.array([2, 3, 4])

a

array([2, 3, 4])

In [274]:
m / a

array([[1., 1., 1.],
       [2., 2., 2.],
       [4., 3., 4.]])

# Ejercicio 147

Mover elementos de un arreglo a otro por medio de índices.

## Solución:

In [275]:
import numpy as np

a = np.array([1, 1, 2, 3, 3])
a

array([1, 1, 2, 3, 3])

In [276]:
b = np.array([0, 4, 6])
b

array([0, 4, 6])

In [277]:
a.put([0, 4], b)

a

array([0, 1, 2, 3, 4])